lib/Model/NameBuilder.cpp

```diff
@@ -119,292 +444,261 @@ llvm::Error model::NameBuilder::populateGlobalNamespace() {
   // Verify needs to verify that each namespace has no internal clashes.
   // Also, the global namespace clashes with everything.
   //
-  // TODO: find a way to add all the helper names to the Result.
+  // TODO: find a way to add all the helper names into the global namespace.
```

Comment:
I think we should prefix them and ban the prefix from names.

Reply:
If we're ok with making them longer, sure. Prefix ideas? `helper_`?

Reply:
Maybe I'd postpone this. It was more a comment on your TODO.

`helper_` would work but helpers already start with `helper_`, however they sometimes get inlined and the functions they call are technically helpers as well but their name doesn't start with `helper_`.

Reply:
Ah, right, this TODO is for the previous iteration, good catch!

> `helper_` would work but helpers already start with `helper_`

Ah, wait, then I might be confused about our terminology. I was assuming _every_ function appearing in the llvm module that is _not_ in the model is a helper of some sort. And some of them do not have any noticeable prefixes. These are what I was referring to when I mentioned helpers. What's the go-to name for them then?

Reply:
(kind of unrelated)

Also looking into this, I noticed another one of the things that still have a `_` prefix: https://github.com/revng/revng/blob/develop/lib/TypeNames/LLVMTypeNames.cpp#L126

Do we have a guarantee somewhere that all of these always start with `helper_` like you mentioned above? Because if so, we should just remove that `_` and ban the prefix.

Reply:
> And some of them do not have any noticeable prefixes. These are what I was referring to when I mentioned helpers. What's the go-to name for them then?

We don't have a very specific terminology there. `:/`
I think helper is fine and we can call those I was talking about "QEMU helpers".

> things that still have a _ prefix

That needs to go.

> Do we have a guarantee somewhere that all of these always start with helper_ like you mentioned above? Because if so, we should just remove that _ and ban the prefix.

No, let's leave a TODO. It's something we have to figure out.

Reply:
Ok, I investigated this a bit and it indeed is awkward. To be on the safe side, I added `helper_` (so we end up with `helper_helper_get_cp_reg64` and friends in some cases) but a lot of others seem fine, like `helper__init_lr` and `helper_llvm_fshl_i32`.

Reply:
Actually, it just hit me while typing this that there's nothing preventing us from doing "add `helper_` only if the name doesn't already start with it 🧠". This way we get ok-looking result for now. And eventually we'll handle this gracefully.

Reply:
Eh, nothing except that you can't link against `libtinycode.ll`. It's not a major use case right now, but I wouldn't rename function arbitrarily.

Is there something preventing us from banning function names that are used in helpers entirely and just emit default names?

I think that'd be the most reasonable thing to do for now.

Reply:
Ah right... Considering all our usages went through the same helper I assumed it was fine. But yeah, external usage of this is also a pain :sadparrot:.

> Is there something preventing us from banning function names that are used in helpers entirely and just emit default names?

We'd need a way to get a list. Preferably without depending on a specific module (That's a pretty heavy dependency to pull to _all the_ places where we emit names - suddenly emitting assembly now requires it - even though it doesn't use helpers).

Reply:
Wait, I wonder how hard it would be to just add wrappers for the offenders 🤔.

I'm thinking a very basic pass that for every helper that doesn't start with `helper_` adds a correctly named function that just calls the `helper_`-less one.

Reply:
This makes me thinkg: maybe `NameBuilder` cannot work "in the void" but should emit different names depending on the artifact. We can guarantee coherence within the same artifact but not across artifacts.

`NameBuilder` could 1) become C agnostic and 2) accept in the constructor a set of reserved names in a specific context (i.e., artifact).
In principle, we could even configure the list of characters allowed in identifiers. For instance, when you're emitting x86-64 you can have `@` in identifier names (I think) but you can't use `rdi` as a function name.

Pro: `NameBuilder` is no longer tied to C.
Cons: assembly and C functions might have different names.

What do you think?